File: recommendation-service/app/middleware/metrics.py
"""
Middleware для сбора метрик Prometheus: число запросов и длительность по endpoint.
"""
import re
import time
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from prometheus_client import Counter, Histogram, REGISTRY

# Нормализуем путь: сегменты из цифр заменяем на :id, чтобы не раздувать метрики
def _normalize_path(path: str) -> str:
    if not path:
        return path
    return re.sub(r"/\d+", "/:id", path)


_requests_total = Counter(
    "recommendation_http_requests_total",
    "Total HTTP requests",
    ["method", "path", "status"],
    registry=REGISTRY,
)
_request_duration_seconds = Histogram(
    "recommendation_http_request_duration_seconds",
    "HTTP request duration in seconds",
    ["method", "path"],
    buckets=(0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0),
    registry=REGISTRY,
)


class PrometheusMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        start = time.perf_counter()
        response = await call_next(request)
        duration = time.perf_counter() - start
        path = _normalize_path(request.url.path)
        method = request.method
        status = str(response.status_code)
        _requests_total.labels(method=method, path=path, status=status).inc()
        _request_duration_seconds.labels(method=method, path=path).observe(duration)
        return response

File: backend/internal/middleware/metrics.go
package middleware

import (
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	httpRequestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "http_requests_total",
			Help: "Total number of HTTP requests",
		},
		[]string{"method", "path", "status"},
	)
	httpRequestDurationSeconds = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "http_request_duration_seconds",
			Help:    "HTTP request duration in seconds",
			Buckets: prometheus.DefBuckets,
		},
		[]string{"method", "path"},
	)
)

// MetricsMiddleware записывает количество запросов и длительность по методу и пути (pattern).
// Путь берётся из c.FullPath() (например /api/v1/users/:id), чтобы не раздувать метрики по ID.
func MetricsMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()
		c.Next()
		duration := time.Since(start).Seconds()
		status := strconv.Itoa(c.Writer.Status())
		path := c.FullPath()
		if path == "" {
			path = c.Request.URL.Path
		}
		method := c.Request.Method
		httpRequestsTotal.WithLabelValues(method, path, status).Inc()
		httpRequestDurationSeconds.WithLabelValues(method, path).Observe(duration)
	}
}
